File: scanner.cpp
#include <iostream>
#include <cstring>
#include <fstream>
#include "token.h"
#include "scanner.h"

using namespace std;

// -----------------------------
// Constructor
// -----------------------------
Scanner::Scanner(const char* s): input(s), first(0), current(0), linea(1), columna(1) {
    }

// -----------------------------
// Funciones auxiliares
// -----------------------------

bool is_white_space(char c) {
    return c == ' ' || c == '\n' || c == '\r' || c == '\t';
}

// SQL no distingue mayusculas en las palabras reservadas, asi que el
// lexema se normaliza antes de compararlo
string a_mayusculas(const string& s) {
    string r = s;
    for (int i = 0; i < (int)r.length(); i++)
        r[i] = toupper(r[i]);
    return r;
}

// consume un caracter. Es el unico lugar donde se mueve current, para que
// linea y columna nunca queden desincronizadas
void Scanner::avanzar() {
    if (input[current] == '\n') {
        linea++;
        columna = 1;
    }
    else {
        columna++;
    }
    current++;
}

// salta espacios en blanco y comentarios de linea (--)
void Scanner::saltarEspacios() {
    while (current < (int)input.length()) {
        if (is_white_space(input[current])) {
            avanzar();
        }
        else if (input[current] == '-' && current + 1 < (int)input.length()
                 && input[current + 1] == '-') {
            while (current < (int)input.length() && input[current] != '\n')
                avanzar();
        }
        else break;
    }
}

// -----------------------------
// nextToken: obtiene el siguiente token
// -----------------------------

Token* Scanner::nextToken() {
    Token* token;

    saltarEspacios();

    // Posicion donde empieza el token, antes de consumir nada
    int linea_tok = linea;
    int columna_tok = columna;

    // Fin de la entrada
    if (current >= (int)input.length()) {
        token = new Token(Token::END);
        token->linea = linea_tok;
        token->columna = columna_tok;
        return token;
    }

    char c = input[current];

    first = current;

    // Numeros (enteros y reales)
    if (isdigit(c)) {
        avanzar();
        while (current < (int)input.length() && isdigit(input[current]))
            avanzar();
        // El punto solo es decimal si le sigue un digito; asi no se
        // confunde 1.5 con la notacion tabla.columna
        if (current + 1 < (int)input.length() && input[current] == '.'
            && isdigit(input[current + 1])) {
            avanzar();
            while (current < (int)input.length() && isdigit(input[current]))
                avanzar();
        }
        token = new Token(Token::NUM, input, first, current - first);
    }
    // Palabras reservadas e identificadores
    else if (isalpha(c) || c == '_') {
        avanzar();
        while (current < (int)input.length()
               && (isalnum(input[current]) || input[current] == '_'))
            avanzar();
        string lexema = a_mayusculas(input.substr(first, current - first));

        // DDL
        if (lexema=="CREATE") token = new Token(Token::CREATE, input, first, current - first);
        else if (lexema=="TABLE") token = new Token(Token::TABLE, input, first, current - first);
        else if (lexema=="INDEX") token = new Token(Token::INDEX, input, first, current - first);
        else if (lexema=="ON") token = new Token(Token::ON, input, first, current - first);
        else if (lexema=="USING") token = new Token(Token::USING, input, first, current - first);
        else if (lexema=="CLUSTERED") token = new Token(Token::CLUSTERED, input, first, current - first);
        else if (lexema=="PRIMARY") token = new Token(Token::PRIMARY, input, first, current - first);
        else if (lexema=="KEY") token = new Token(Token::KEY, input, first, current - first);

        // porganizacion del archivo (2.1.1)
        else if (lexema=="HEAP") token = new Token(Token::HEAP, input, first, current - first);
        else if (lexema=="SEQUENTIAL") token = new Token(Token::SEQUENTIAL, input, first, current - first);

        // tipos de indice (2.1.2)
        else if (lexema=="BTREE") token = new Token(Token::BTREE, input, first, current - first);
        else if (lexema=="HASH") token = new Token(Token::HASH, input, first, current - first);

        // tipos de dato
        else if (lexema=="INT") token = new Token(Token::INT, input, first, current - first);
        else if (lexema=="FLOAT") token = new Token(Token::FLOAT, input, first, current - first);
        else if (lexema=="BOOL") token = new Token(Token::BOOL, input, first, current - first);
        else if (lexema=="DATE") token = new Token(Token::DATE, input, first, current - first);
        else if (lexema=="VARCHAR") token = new Token(Token::VARCHAR, input, first, current - first);

        //dml
        else if (lexema=="SELECT") token = new Token(Token::SELECT, input, first, current - first);
        else if (lexema=="FROM") token = new Token(Token::FROM, input, first, current - first);
        else if (lexema=="JOIN") token = new Token(Token::JOIN, input, first, current - first);
        else if (lexema=="WHERE") token = new Token(Token::WHERE, input, first, current - first);
        else if (lexema=="GROUP") token = new Token(Token::GROUP, input, first, current - first);
        else if (lexema=="ORDER") token = new Token(Token::ORDER, input, first, current - first);
        else if (lexema=="BY") token = new Token(Token::BY, input, first, current - first);
        else if (lexema=="ASC") token = new Token(Token::ASC, input, first, current - first);
        else if (lexema=="DESC") token = new Token(Token::DESC, input, first, current - first);
        else if (lexema=="LIMIT") token = new Token(Token::LIMIT, input, first, current - first);
        else if (lexema=="INSERT") token = new Token(Token::INSERT, input, first, current - first);
        else if (lexema=="INTO") token = new Token(Token::INTO, input, first, current - first);
        else if (lexema=="VALUES") token = new Token(Token::VALUES, input, first, current - first);
        else if (lexema=="DELETE") token = new Token(Token::DELETE, input, first, current - first);

        // transacciones (2.1.4)
        else if (lexema=="BEGIN") token = new Token(Token::BEGIN, input, first, current - first);
        else if (lexema=="END") token = new Token(Token::END_KW, input, first, current - first);
        else if (lexema=="TRANSACTION") token = new Token(Token::TRANSACTION, input, first, current - first);

        // condiciones
        else if (lexema=="AND") token = new Token(Token::AND, input, first, current - first);
        else if (lexema=="OR") token = new Token(Token::OR, input, first, current - first);
        else if (lexema=="BETWEEN") token = new Token(Token::BETWEEN, input, first, current - first);
        else if (lexema=="TRUE") token = new Token(Token::TRUE_KW, input, first, current - first);
        else if (lexema=="FALSE") token = new Token(Token::FALSE_KW, input, first, current - first);

        // funciones de agregacoin
        else if (lexema=="COUNT") token = new Token(Token::COUNT, input, first, current - first);
        else if (lexema=="SUM") token = new Token(Token::SUM, input, first, current - first);
        else if (lexema=="AVG") token = new Token(Token::AVG, input, first, current - first);
        else if (lexema=="MIN") token = new Token(Token::MIN, input, first, current - first);
        else if (lexema=="MAX") token = new Token(Token::MAX, input, first, current - first);

        else token = new Token(Token::ID, input, first, current - first);
    }
    // Cadenas entre comillas simples ('' representa una comilla literal)
    else if (c == '\'') {
        avanzar();
        string texto = "";
        bool cerrada = false;
        while (current < (int)input.length() && input[current] != '\n') {
            if (input[current] == '\'') {
                if (current + 1 < (int)input.length() && input[current+1] == '\'') {
                    texto += '\'';
                    avanzar();
                    avanzar();
                    continue;
                }
                avanzar();
                cerrada = true;
                break;
            }
            texto += input[current];
            avanzar();
        }
        if (!cerrada) {
            token = new Token(Token::ERR, '\'');
        }
        else {
            token = new Token(Token::STR);
            token->text = texto;
        }
    }
    // operadores y signos de puntuacion
    else if (strchr(",;().*=<>!", c)) {
        switch (c) {
            case ',': token = new Token(Token::COMA,    c); break;
            case ';': token = new Token(Token::SEMICOL, c); break;
            case '(': token = new Token(Token::LPAREN,  c); break;
            case ')': token = new Token(Token::RPAREN,  c); break;
            case '.': token = new Token(Token::PUNTO,   c); break;
            case '*': token = new Token(Token::STAR,    c); break;
            case '=': token = new Token(Token::EQ,      c); break;
            case '<':
            if (input[current+1]=='=')
            {
                avanzar();
                token = new Token(Token::LE, input, first, current + 1 - first);
            }
            else if (input[current+1]=='>')
            {
                avanzar();
                token = new Token(Token::NEQ, input, first, current + 1 - first);
            }
            else{
                token = new Token(Token::LT, c);
            }
            break;
            case '>':
            if (input[current+1]=='=')
            {
                avanzar();
                token = new Token(Token::GE, input, first, current + 1 - first);
            }
            else{
                token = new Token(Token::GT, c);
            }
            break;
            case '!':
            if (input[current+1]=='=')
            {
                avanzar();
                token = new Token(Token::NEQ, input, first, current + 1 - first);
            }
            else{
                token = new Token(Token::ERR, c);
            }
            break;
            default: token = new Token(Token::ERR, c); break;
        }
        avanzar();
    }

    // Caracter invalido
    else {
        token = new Token(Token::ERR, c);
        avanzar();
    }

    // Estampar la posicion donde empezo el token
    token->linea = linea_tok;
    token->columna = columna_tok;
    return token;
}

// -----------------------------
// Destructor
// -----------------------------
Scanner::~Scanner() { }

// -----------------------------
// Funcion de prueba
// -----------------------------

void ejecutar_scanner(Scanner* scanner, const string& InputFile) {
    Token* tok;

    // Crear nombre para archivo de salida
    string OutputFileName = InputFile;
    size_t pos = OutputFileName.find_last_of(".");
    if (pos != string::npos) {
        OutputFileName = OutputFileName.substr(0, pos);
    }
    OutputFileName += "_tokens.txt";

    ofstream outFile(OutputFileName);
    if (!outFile.is_open()) {
        cerr << "Error: no se pudo abrir el archivo " << OutputFileName << endl;
        return;
    }

    outFile << "Scanner\n" << endl;

    while (true) {
        tok = scanner->nextToken();

        if (tok->type == Token::END) {
            outFile << *tok << endl;
            delete tok;
            outFile << "\nScanner exitoso" << endl << endl;
            outFile.close();
            return;
        }

        if (tok->type == Token::ERR) {
            outFile << *tok << endl;
            outFile << "Caracter invalido en linea " << tok->linea
                    << ", columna " << tok->columna << endl << endl;
            delete tok;
            outFile << "Scanner no exitoso" << endl << endl;
            outFile.close();
            return;
        }

        outFile << *tok << endl;
        delete tok;
    }
}

File: parser/scanner.py
from token_sql import Token


# -----------------------------
# Funciones auxiliares
# -----------------------------

def is_white_space(c):
    return c == ' ' or c == '\n' or c == '\r' or c == '\t'


# SQL no distingue mayusculas en las palabras reservadas, asi que el
# lexema se normaliza antes de compararlo
def a_mayusculas(s):
    return s.upper()


# Tabla de palabras reservadas
PALABRAS_RESERVADAS = {
    # DDL
    "CREATE": Token.Type.CREATE,
    "TABLE": Token.Type.TABLE,
    "INDEX": Token.Type.INDEX,
    "ON": Token.Type.ON,
    "USING": Token.Type.USING,
    "CLUSTERED": Token.Type.CLUSTERED,
    "PRIMARY": Token.Type.PRIMARY,
    "KEY": Token.Type.KEY,

    # Organizacion del archivo (2.1.1)
    "HEAP": Token.Type.HEAP,
    "SEQUENTIAL": Token.Type.SEQUENTIAL,

    # Tipos de indice (2.1.2)
    "BTREE": Token.Type.BTREE,
    "HASH": Token.Type.HASH,

    # Tipos de dato
    "INT": Token.Type.INT,
    "FLOAT": Token.Type.FLOAT,
    "BOOL": Token.Type.BOOL,
    "DATE": Token.Type.DATE,
    "VARCHAR": Token.Type.VARCHAR,

    # DML
    "SELECT": Token.Type.SELECT,
    "FROM": Token.Type.FROM,
    "JOIN": Token.Type.JOIN,
    "WHERE": Token.Type.WHERE,
    "GROUP": Token.Type.GROUP,
    "ORDER": Token.Type.ORDER,
    "BY": Token.Type.BY,
    "ASC": Token.Type.ASC,
    "DESC": Token.Type.DESC,
    "LIMIT": Token.Type.LIMIT,
    "INSERT": Token.Type.INSERT,
    "INTO": Token.Type.INTO,
    "VALUES": Token.Type.VALUES,
    "DELETE": Token.Type.DELETE,

    # Transacciones (2.1.4)
    "BEGIN": Token.Type.BEGIN,
    "END": Token.Type.END_KW,
    "TRANSACTION": Token.Type.TRANSACTION,

    # Condiciones
    "AND": Token.Type.AND,
    "OR": Token.Type.OR,
    "BETWEEN": Token.Type.BETWEEN,
    "TRUE": Token.Type.TRUE_KW,
    "FALSE": Token.Type.FALSE_KW,

    # Funciones de agregacion
    "COUNT": Token.Type.COUNT,
    "SUM": Token.Type.SUM,
    "AVG": Token.Type.AVG,
    "MIN": Token.Type.MIN,
    "MAX": Token.Type.MAX,
}


class Scanner:

    # Constructor
    def __init__(self, in_s):
        self.input = in_s
        self.first = 0
        self.current = 0
        self.linea = 1      # Fila actual (1-based)
        self.columna = 1    # Columna actual (1-based)

    # Consume un caracter. Es el unico lugar donde se mueve current, para
    # que linea y columna nunca queden desincronizadas
    def avanzar(self):
        if self.input[self.current] == '\n':
            self.linea += 1
            self.columna = 1
        else:
            self.columna += 1
        self.current += 1

    # Salta espacios en blanco y comentarios de linea (--)
    def saltar_espacios(self):
        while self.current < len(self.input):
            if is_white_space(self.input[self.current]):
                self.avanzar()
            elif (self.input[self.current] == '-'
                  and self.current + 1 < len(self.input)
                  and self.input[self.current + 1] == '-'):
                while (self.current < len(self.input)
                       and self.input[self.current] != '\n'):
                    self.avanzar()
            else:
                break

    # -----------------------------
    # next_token: obtiene el siguiente token
    # -----------------------------
    def next_token(self):
        self.saltar_espacios()

        # Posicion donde empieza el token, antes de consumir nada
        linea_tok = self.linea
        columna_tok = self.columna

        # Fin de la entrada
        if self.current >= len(self.input):
            return Token(Token.Type.END, "", linea_tok, columna_tok)

        c = self.input[self.current]

        self.first = self.current

        # Numeros (enteros y reales)
        if c.isdigit():
            self.avanzar()
            while self.current < len(self.input) and self.input[self.current].isdigit():
                self.avanzar()
            # El punto solo es decimal si le sigue un digito; asi no se
            # confunde 1.5 con la notacion tabla.columna
            if (self.current + 1 < len(self.input)
                    and self.input[self.current] == '.'
                    and self.input[self.current + 1].isdigit()):
                self.avanzar()
                while self.current < len(self.input) and self.input[self.current].isdigit():
                    self.avanzar()
            token = Token(Token.Type.NUM,
                          self.input[self.first:self.current])

        # Palabras reservadas e identificadores
        elif c.isalpha() or c == '_':
            self.avanzar()
            while (self.current < len(self.input)
                   and (self.input[self.current].isalnum()
                        or self.input[self.current] == '_')):
                self.avanzar()
            texto = self.input[self.first:self.current]
            lexema = a_mayusculas(texto)

            if lexema in PALABRAS_RESERVADAS:
                token = Token(PALABRAS_RESERVADAS[lexema], texto)
            else:
                token = Token(Token.Type.ID, texto)

        # Cadenas entre comillas simples ('' representa una comilla literal)
        elif c == '\'':
            self.avanzar()
            texto = ""
            cerrada = False
            while self.current < len(self.input) and self.input[self.current] != '\n':
                if self.input[self.current] == '\'':
                    if (self.current + 1 < len(self.input)
                            and self.input[self.current + 1] == '\''):
                        texto += '\''
                        self.avanzar()
                        self.avanzar()
                        continue
                    self.avanzar()
                    cerrada = True
                    break
                texto += self.input[self.current]
                self.avanzar()
            if not cerrada:
                token = Token(Token.Type.ERR, '\'')
            else:
                token = Token(Token.Type.STR, texto)

        # Operadores y signos de puntuacion
        elif c in ",;().*=<>!":
            if c == ',':
                token = Token(Token.Type.COMA, c)
            elif c == ';':
                token = Token(Token.Type.SEMICOL, c)
            elif c == '(':
                token = Token(Token.Type.LPAREN, c)
            elif c == ')':
                token = Token(Token.Type.RPAREN, c)
            elif c == '.':
                token = Token(Token.Type.PUNTO, c)
            elif c == '*':
                token = Token(Token.Type.STAR, c)
            elif c == '=':
                token = Token(Token.Type.EQ, c)
            elif c == '<':
                if self.sigue('='):
                    self.avanzar()
                    token = Token(Token.Type.LE, "<=")
                elif self.sigue('>'):
                    self.avanzar()
                    token = Token(Token.Type.NEQ, "<>")
                else:
                    token = Token(Token.Type.LT, c)
            elif c == '>':
                if self.sigue('='):
                    self.avanzar()
                    token = Token(Token.Type.GE, ">=")
                else:
                    token = Token(Token.Type.GT, c)
            elif c == '!':
                if self.sigue('='):
                    self.avanzar()
                    token = Token(Token.Type.NEQ, "!=")
                else:
                    token = Token(Token.Type.ERR, c)
            self.avanzar()

        # Caracter invalido
        else:
            token = Token(Token.Type.ERR, c)
            self.avanzar()

        # Estampar la posicion donde empezo el token
        token.linea = linea_tok
        token.columna = columna_tok
        return token

    # Comprueba si el caracter siguiente al actual es el esperado
    def sigue(self, c):
        return (self.current + 1 < len(self.input)
                and self.input[self.current + 1] == c)


# -----------------------------
# Funcion de prueba
# -----------------------------

def ejecutar_scanner(scanner, input_file):
    # Crear nombre para archivo de salida
    pos = input_file.rfind(".")
    if pos != -1:
        output_file_name = input_file[:pos]
    else:
        output_file_name = input_file
    output_file_name += "_tokens.txt"

    try:
        out_file = open(output_file_name, "w", encoding="utf-8")
    except OSError:
        print("Error: no se pudo abrir el archivo " + output_file_name)
        return

    out_file.write("Scanner\n\n")

    while True:
        tok = scanner.next_token()

        if tok.type == Token.Type.END:
            out_file.write(str(tok) + "\n")
            out_file.write("\nScanner exitoso\n\n")
            out_file.close()
            return

        if tok.type == Token.Type.ERR:
            out_file.write(str(tok) + "\n")
            out_file.write("Caracter invalido en linea " + str(tok.linea)
                           + ", columna " + str(tok.columna) + "\n\n")
            out_file.write("Scanner no exitoso\n\n")
            out_file.close()
            return

        out_file.write(str(tok) + "\n")
